### PR TITLE
[internal/exp/metrics] Add a new internal package for handling metric staleness

### PR DESCRIPTION
**Pull request by RichieSams**
**Description:**
It's a glorified wrapper over a Map type, which allows values to be expired based on a pre-supplied interval.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31016

**Testing:**
I added some basic tests of the PriorityQueue implementation as well as the expiry behaviour of Staleness

**Documentation:**

All the new structs are documented

### CODE REVIEW

File: internal/exp/metrics/staleness/staleness.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// We override how Now() is returned, so we can have deterministic tests
+var nowFunc = time.Now
+
+type Map[T any] interface {
+	Load(key identity.Stream) (T, bool)
+	Store(key identity.Stream, value T)
+	Delete(key identity.Stream)
+	// Range calls f sequentially for each key and value present in the map.
+	// If f returns false, range stops the iteration.
+	Range(f func(key identity.Stream, value T) bool)
```
Comment (sh0rez) [inline internal/exp/metrics/staleness/staleness.go:21]: I think here is an opportunity to already conform to the future rangefunc api: https://go.dev/wiki/RangefuncExperiment

```go
func (m RawMap[T]) Items() iter.Seq2[identity.Stream, T] {
    return func(yield func(identity.Stream, T) bool) bool {
        for k, v := range m {
            if !yield(k, v) { break }
        }
        return false
    }
}
```

File: internal/exp/metrics/staleness/staleness_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+)
+
+// RawMap
+var _ Map[time.Time] = (*RawMap[identity.Stream, time.Time])(nil)
+
+type RawMap[K comparable, V any] map[K]V
+
+func (rm *RawMap[K, V]) Load(key K) (V, bool) {
+	value, ok := (*rm)[key]
```
Comment (sh0rez) [inline internal/exp/metrics/staleness/staleness_test.go:18]: given the cache-like nature, I think callers would benefit from having `Load()` initialize to zero-value when not-exist, eliminating the need for a load-check-store

File: internal/exp/metrics/staleness/priority_queue.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}
+}
+
+func (pq *PriorityQueue) Peek() (identity.Stream, time.Time) {
+	val := pq.inner[0]
+	return val.key, val.prio
+}
+
+func (pq *PriorityQueue) Pop() (identity.Stream, time.Time) {
+	val := heap.Pop(&pq.inner).(*queueItem)
```
Comment (sh0rez) [inline internal/exp/metrics/staleness/priority_queue.go:91]: does pop also have to delete the item from the lookup map, so it doesn't hold onto every historical item pointer indefinitely?

File: internal/exp/metrics/staleness/priority_queue.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	*pq = old[0 : n-1]
+	return item
+}
+
+func (pq *priorityQueueImpl) Update(item *queueItem, newPrio time.Time) {
+	item.prio = newPrio
+	heap.Fix(pq, item.index)
+}
+
+type PriorityQueue struct {
```
Comment (sh0rez) [inline internal/exp/metrics/staleness/priority_queue.go:56]: wdyt about `PriorityQueue` (as used in the Staleness type) being an interface and having something like private like `heapQueue` be the actual implementation. Couples those two a little less

**Review by tombrk — CHANGES_REQUESTED** [review]
This is really nice and exactly what I had in mind. Great work!!

File: internal/exp/metrics/staleness/staleness.go (hunk 1 above)
Comment (RichieSams) [reply internal/exp/metrics/staleness/staleness.go:21]: That said, it looks like for `iter.Seq2` to be defined, we need go 1.22 and `GOEXPERIMENT=rangefunc`. Which I don't think we can allow for all the contrib.

Should we implement that ourselves for now? Or stick with Range until iter.Seq2 is GA?

File: internal/exp/metrics/staleness/staleness_test.go (hunk 2 above)
Comment (RichieSams) [reply internal/exp/metrics/staleness/staleness_test.go:18]: We could add `LoadOrStore(key K, value V) (V, bool)`. Which would do what you're saying.

Comment (RichieSams) [reply internal/exp/metrics/staleness/staleness_test.go:18]: Added

File: internal/exp/metrics/staleness/priority_queue.go (hunk 3 above)
Comment (RichieSams) [reply internal/exp/metrics/staleness/priority_queue.go:98]: Fixed

File: internal/exp/metrics/staleness/priority_queue.go (hunk 4 above)
Comment (RichieSams) [reply internal/exp/metrics/staleness/priority_queue.go]: Done

File: internal/exp/metrics/staleness/staleness.go (hunk 1 above)
Comment (sh0rez) [reply internal/exp/metrics/staleness/staleness.go:27]: iter.Seq2 is merely an alias for `func(yield func(K, V) bool) bool`, so we can also just this raw type in the signature, which will be forward compatible!

Comment (RichieSams) [reply internal/exp/metrics/staleness/staleness.go:27]: Ah good point. Thanks!

File: internal/exp/metrics/staleness/staleness.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+type Map[T any] interface {
+	// Load the value at key. If it does not exist, the boolean will be false and the value returned will be the zero value
+	Load(key identity.Stream) (T, bool)
+	// Store the given key value pair in the map
+	Store(key identity.Stream, value T)
+	// LoadOrStore will either load the value from the map and return it and the boolean `true`
+	// or if it doesn't exist in the Map yet, the value passed in will be stored and then returned with the boolean `false`
+	LoadOrStore(key identity.Stream, value T) (T, bool)
+	// Remove the value at key from the map
```
Comment (sh0rez) [inline internal/exp/metrics/staleness/staleness.go:23]: would like to revert my previous opinion, this is actually a lot less necessary than I thought.
lets remove it to keep the api surface smaller ;)

**Review by tombrk — COMMENTED** [review]
tried this out, works really well!
I think this is ready for final review once rebased onto HEAD!

**Comment by RichieSams** [comment]
@jpkrohling @fatsheep9146 @bryan-aguilar 

This PR is ready for final review. (Can one of you add the "Skip Changelog" label, since this is an internal-only change)

@sh0rez Any final comments?

File: internal/exp/metrics/staleness/map.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	// See: https://go.dev/wiki/RangefuncExperiment
+	Items() func(yield func(identity.Stream, T) bool) bool
+}
+
+// RawMap implementation
+
+var _ Map[time.Time] = (*RawMap[identity.Stream, time.Time])(nil)
+
+// RawMap is an implementation of the Map interface using a standard golang map
+type RawMap[K comparable, V any] map[K]V
```
Comment (sh0rez) [inline internal/exp/metrics/staleness/map.go:30]: I think this type doesn't creeps a bit beyond the scope of this package.
I like that this package defines the `Map[T]` interface as the behavior it expects to fulfill its purpose.
A general stream map would feel more fitting as a `streams.HashMap` or similar.

I'd be okay with this pkg not defining a map implementation at all, leaving that to consumers.
No hard opinion however, if you prefer we can leave it in here for now, as we'll likely iterate on these packages for a bit anyways.

**Review by tombrk — COMMENTED** [review]
I think it's in great shape for the initial bring-up. Everything else can be iterated upon :)

File: internal/exp/metrics/staleness/map.go (hunk 6 above)
Comment (RichieSams) [reply internal/exp/metrics/staleness/map.go:30]: > I'd be okay with this pkg not defining a map implementation at all, leaving that to consumers.

That was my initial plan. I had implemented RawMap just in the tests. But then I realized that I would need to re-implement the same code in the interval processor (and perhaps you as well in deltatocumulative). So I moved it to be in the package proper.

But I'm kind of the same. I don't feel strongly either way

**Review by djaglowski — COMMENTED** [review]
LGTM but I will wait to approve until @sh0rez's feedback is fully addressed.

File: internal/exp/metrics/staleness/map.go (hunk 6 above)
Comment (RichieSams) [reply internal/exp/metrics/staleness/map.go:30]: Are you ok with leaving it in for now? And iterating as the various processors utilize it?

Comment (RichieSams) [reply internal/exp/metrics/staleness/map.go:30]: Bump

**Review by tombrk — APPROVED** [review]

**Comment by sh0rez** [comment]
@djaglowski ready to merge from my side!

Think this still needs the "no changelog" label

**Review by djaglowski — APPROVED** [review]

**Pull request merged**